### PR TITLE
PyTorch migration: Remove tensorflow components, add FATE estimators

### PR DESCRIPTION
**Pull request by timokau**
## Description

See [this comment](https://github.com/kiudee/cs-ranking/pull/164#issuecomment-812940505) for a description of the current status.

## Motivation and Context

Tensorflow 1 is deprecated and we need to move away from it. This PR is an attempt to evaluate pytorch as an alternative. ~For now I don't try to fit the existing API (at least not yet).~

## How Has This Been Tested?

Lints & tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by kiudee** [comment]
Already looks very clean - well done!

**Comment by timokau** [comment]
Just as a little status update, since this has been going on for a while: I think this is turning out quite nicely. I have implemented FETA ranking and FETA discrete choice. That shows flexibility on one axis (result type). I plan to also implement the same for FATE to show the flexibility on the other axis and finish the proof of concept. At that point we could evaluate and see where to take it from there.

So in summary, things are moving along but are not quite ready for review/discussion yet. Hopefully soon-ish.

**Comment by timokau** [comment]
Okay, I think this is sufficient as a proof of concept now. I have implemented FATE and FETA, each in the ranking and discrete choice variant.

I replaced a lot of the inheritance in the current tensorflow implementation with composition. I have split the code into "scoring modules" and estimators. The scoring modules are themselves composed of smaller modules, which makes them easier to reuse/understand/test.

I have based the estimator implementation on skorch, which takes care of a lot of the boilerplate for us. We no longer have to care about training loops, instantiating optimizers or passing the parameters to uninitialized classes. We get #116 basically for free.

The actual "heavy lifing" of the computation (the pairwise utilities) is disentangled from the FETA/FATE architecture (the "data flow" part), so its easy to modify or replace. For now its just a simple 1-layer linear network. This decomposition of scorer/estimator/utility removes a lot of duplication. It would be very easy to add a new scorer (for example based on graph neural networks) and "throw" it at the existing Ranking/Discrete choice estimators. It would also be very easy to derive a new utility function architecture and "throw" that at the FATE module.

If you want to look at the implementation, here are the most interesting files:
- `modules/scoring.py` takes care of the high level assembly for the FATE and FETA scoring modules.
- `estimators.py` derives ranking and discrete choice estimators from the scorers.
- `*_losses.py`, `*_datasets.py` defines some losses and test datasets. `modules/feta_support.py` and `modules/embedding.py` are contain the more low-level aspects of FETA and FATE.

What do you think @kiudee? There are still things to improve of course, but I think its sufficient as a proof of concept.

**Comment by timokau** [comment]
Also CC @prithagupta if you are interested in this.

File: poc/modules/scoring.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    >>> scoring(instances)
+    tensor([[ 4.6000,  6.6000,  8.6000],
+            [16.2000, 18.2000, 20.2000]])
+    """
+
+    # TODO use a more powerful pairwise utility module
+    def __init__(self, n_features, pairwise_utility_module=PairwiseLinearUtility):
+        super().__init__()
+        self.mean_aggregated_utilty = MeanAggregatedUtility(
+            FirstOrderUtility(pairwise_utility_module(n_features))
```
Comment (kiudee) [inline poc/modules/scoring.py:53]: I like the composability with different pairwise models here.

File: poc/modules/scoring.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        n_objects = instances.size(1)
+        contexts = self.embedding(instances)
+        # Repeat each context for each object within the instance; This is then
+        # a flat list of contexts. Then reshape to have a list of contexts per
+        # instance.
+        context_per_object = contexts.repeat_interleave(n_objects, dim=0).reshape_as(
+            instances
+        )
+        pairs = torch.cat((instances, context_per_object), dim=-1)
+        utilities = self.pairwise_utility_module(pairs)
```
Comment (kiudee) [inline poc/modules/scoring.py:145]: Very clean decomposition.

File: poc/modules/scoring.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    >>> scoring(instances)
+    tensor([[ 4.6000,  6.6000,  8.6000],
+            [16.2000, 18.2000, 20.2000]])
+    """
+
+    # TODO use a more powerful pairwise utility module
+    def __init__(self, n_features, pairwise_utility_module=PairwiseLinearUtility):
+        super().__init__()
+        self.mean_aggregated_utilty = MeanAggregatedUtility(
```
Comment (kiudee) [inline poc/modules/scoring.py:52]: In principle we can even think about making this modular. One could use different aggregation functions here or even a learned aggregation operator.

File: poc/modules/scoring.py (hunk 2 above)
Comment (timokau) [reply poc/modules/scoring.py:145]: Thank you!

File: poc/modules/scoring.py (hunk 3 above)
Comment (timokau) [reply poc/modules/scoring.py:52]: Yes, I think that would be interesting :) The composition architecture should make experiments like that much easier.

**Comment by timokau** [comment]
What is your general verdict @kiudee? Should I continue down this path, implementing more of the existing learners and functionality and eventually replacing the current implementation? Or rather try something else?

File: poc/datasets/variable_choice_datasets.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            # center. Center is some point on the unit ball (cluster_spread).
+            x, y = make_randn_pareto_choices(
+                cluster_size=cluster_size,
+                n_features=n_features,
+                n_objects=n_objects,
+                center=center,
+            )
+            # TODO(next) how do clusters and instances relate? Is each cluster
+            # one instance? Then why the n_instances / cluster_size in the for
+            # loop?
```
Comment (timokau) [inline poc/datasets/variable_choice_datasets.py:193]: @kiudee could you help me understand how the Pareto dataset is supposed to work? Intuitively, I would just generate one "cluster" per instance and then try to predict the pareto front of this cluster.

If I understand the cs-ranking code correctly however, it generates clusters with a hard-coded size of 10 and then splits those clusters along the instances. So a single problem-instance might contain points from two different clusters. Is that correct? Is that intended? If so, is the model supposed to perform clustering as well as determining the Pareto-front?

Comment (kiudee) [reply poc/datasets/variable_choice_datasets.py:193]: Correct, for each instance you generate a centroid and sample points from the Gaussian centered on that centroid.
If the current code does something different, then it is definitely not intended.

Comment (timokau) [reply poc/datasets/variable_choice_datasets.py:193]: Okay, I read the code again and it seems to work slightly differently then I originally thought, but it also doesn't generate one cluster+centroid per instance. I think I was a bit confused by the parameter names.

What it actually does is
1. generate a centroid (this is the `sample_unit_ball` function, which is somewhat confusing since it is only ever used to sample a single point)
2. generate n_objects random objects; shift them to the centroid; compute the pareto front (this is `make_randn_pareto_choices`)
3. repeat this 10 times (the `cluster_size` parameter, defaults to 10 and is never overriden) **for the same centroid** (the `for` loop in `make_globular_pareto_choices` and the `n_instances` parameter to `make_randn_pareto_choices` which **does not** refer to the actual number of problem instances here)
4. return to step 1

Since it rounds the number of generated instances to a multiple of 10, you get uninitialized values when you request something that is not a multiple of 10:

```
[ins] In [1]: from csrank import ChoiceDatasetGenerator
[ins] In [2]: gen = ChoiceDatasetGenerator(
         ...:     dataset_type="pareto",
         ...:     random_state=42,
         ...:     n_train_instances=11,
         ...:     n_test_instances=1,
         ...:     n_objects=2,
         ...:     n_features=2,
         ...: )
[ins] In [3]: X_train, Y_train, X_test, Y_test = gen.get_single_train_test_split()
[ins] In [4]: Y_train
Out[5]: 
array([[                  1,                   1],
       [                  1,                   1],
       [                  1,                   0],
       [                  0,                   1],
       [                  0,                   1],
       [                  0,                   1],
       [                  0,                   1],
       [                  0,                   1],
       [                  1,                   1],
       [7737013110153509221, 7308039772162521445],
       [                  1,                   1]])
```

At least the last part is clearly a bug. The rest seems very odd, but also intended -- its hard to imagine it was implemented this way by accident. What is going on here? Should I change it to generate a unique centroid per problem instance?

Comment (timokau) [reply poc/datasets/variable_choice_datasets.py:193]: I'm also having trouble understanding *why* the cluster centroid is generated the way it is. Here is the function:

```python
def sample_unit_ball(n_inst=10000, n_features=2, rng=None, radius=1.0):
    rng = check_random_state(rng)
    X = rng.randn(n_inst, n_features)
    u = rng.uniform(size=n_inst)[:, None]
    X /= np.linalg.norm(X, axis=1, ord=2)[:, None]
    X *= radius * u
    return X
```

where `n_inst` is always 1, so we can ignore that. Why do we first sample from a normal distribution, then additionally from a uniform distribution? Doesn't that just arbitrarily bias the sample towards 0? What is the np.linalg.norm for?

Naively, I would just write `np.random.randn(n_features) * (radius ** 1/n_features)` to generate the centroid (if I wanted to draw it from a normal distribution -- the same with `np.random.uniform` otherwise).

File: poc/datasets/variable_choice_datasets.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    for i, attr in enumerate(X):
+        # A binary flag array that has an entry for each other point: 1
+        # if and only if it is "better" than the other point in at
+        # least one dimension (i.e. it is not dominated by that point).
+        not_dominated = np.any((X * signs[None, :]) <= (attr * signs), axis=1)
+        pareto[i] = np.all(not_dominated)
+    return pareto
+
+
+class ParetoChoiceProblem(TensorDataset):
```
Comment (timokau) [inline poc/datasets/variable_choice_datasets.py:101]: @kiudee please have a look at this simplified generator. Is that how the problem was intended? I have skipped the radius that the centroids are sampled from, since that was never changed anyway and it doesn't really change the model's task.

File: poc/datasets/variable_choice_datasets.py (hunk 4 above)
Comment (kiudee) [reply poc/datasets/variable_choice_datasets.py]: The goal of `sample_unit_ball` is to sample *uniformly* from the unit ball.
With `rng.randn` we generate a random direction and normalize it. Then we generate a radius uniformly at random. That will ensure that the whole ball is covered and that there is no concentration around the centroid.

Comment (kiudee) [reply poc/datasets/variable_choice_datasets.py]: > At least the last part is clearly a bug. The rest seems very odd, but also intended -- its hard to imagine it was implemented this way by accident. What is going on here? Should I change it to generate a unique centroid per problem instance?

Yes, that sounds like the most sensible thing to do.

File: poc/datasets/variable_choice_datasets.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            pareto_front_flags = _compute_pareto_front(x_instance)
+            x_instances.append(x_instance)
+            y_instances.append(pareto_front_flags)
+        super().__init__(
+            torch.tensor(x_instances).float(), torch.tensor(y_instances).float()
+        )
+
+    def draw_centroid(self):
+        # Draw a centroid from the unit ball from a Gaussian distribution.
+        return self.random_state.randn(self.n_features)
```
Comment (kiudee) [inline poc/datasets/variable_choice_datasets.py:153]: As mentioned above, this would change the distribution of clusters slightly (more concentrated around the centroid). This is not problematic per se, but could make it easier for context-independent models to learn a utility function.

File: poc/datasets/variable_choice_datasets.py (hunk 4 above)
Comment (timokau) [reply poc/datasets/variable_choice_datasets.py]: > With `rng.randn` we generate a random direction and normalize it. Then we generate a radius uniformly at random. That will ensure that the whole ball is covered and that there is no concentration around the centroid.

Ah, that makes a lot of sense. Thank you for the explanation.

> Yes, that sounds like the most sensible thing to do.

Great, I'll prepare a PR.

File: poc/datasets/variable_choice_datasets.py (hunk 6 above)
Comment (timokau) [reply poc/datasets/variable_choice_datasets.py:153]: I see, thank you for your explanation! I'll fix it in the POC implementation and also adapt the current implementation as discussed in the other thread.

File: poc/datasets/variable_choice_datasets.py (hunk 4 above)
Comment (timokau) [reply poc/datasets/variable_choice_datasets.py]: > Great, I'll prepare a PR.

https://github.com/kiudee/cs-ranking/pull/177

**Comment by timokau** [comment]
Another status update: I'm experimenting with experiments. We should be able to reproduce the experiments of the main papers with the new implementation, and I'd like to be able to do that in an easily reproducible way (that could possibly be repeated on each release). I'm trying to use `sacred` for the purpose. I'm abusing the "named configuration" system a bit, but currently you can do things like

```
python3 experiment.py -m sacred with feta_variable_choice_estimator pareto_choice_problem dataset_params.n_instances=10000
```

you can pick "named configurations" for an estimator and a dataset. You can then overwrite all parameters on the command line. Sacred will run the experiment, store everything that is needed to reproduce it and also store the results in a database:
2020-11-24-212138_1918x1058_scrot

**Comment by timokau** [comment]
Some more progress: I've added some metrics and played with the experiments a bit. Here I was trying to see how far I could push the current feta implementation with its defaults and just 1000 pareto instances (which was further than expected):
training plot
I ended up stopping the training even though the informedness still seemed to be rising very slightly. I ran the experiment with

```
python3 poc/experiment.py -m sacred with feta_variable_choice_estimator pareto_choice_problem dataset_params.n_instances=1000 dataset_params.n_objects=30
```

I also created an upstream PR for the Sacred logger for skorch: https://github.com/skorch-dev/skorch/pull/725

File: poc/experiment.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+
+@ex.named_config
+def learning_choice_functions_feta_pareto():
+    """A (partial) reproduction of the experimental setup of the FETA paper."""
+    # Refer to Table 2 in https://arxiv.org/abs/1901.10860 for hyperparameters
+    # and reasonable ranges.
+    estimator = FETAVariableChoiceEstimator
+    estimator_params = {
+        "max_epochs": int(1e9),  # TODO look up when to terminate in the paper
```
Comment (timokau) [inline poc/experiment.py:106]: @kiudee do you recall what stopping criterion you used in the experiments of ["Learning Choice Functions: Concepts and Architectures"](https://arxiv.org/abs/1901.10860)? I can't find that information in the paper. It mentions a tolerance parameter, but that is only used for PairwiseSVM. The `epochs` parameter in the current cs-ranking implementation usually defaults to `10`, which seems very small (it should be at least something in the order of `edrop` in the paper).

Was it a fixed epoch value or some sort of convergence criterion?

Comment (kiudee) [reply poc/experiment.py:106]: As far as I remember we used 1000 epochs. It is true that we did not mention this in the paper.

File: poc/experiment.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    estimator_type = "variable_choice"
+    dataset = ParetoChoiceProblem
+    dataset_params = {
+        "n_features": 2,
+        "n_objects": 30,
+        "n_instances": int(1e5),
+    }
+    dataset_type = "variable_choice"
+    # TODO set cross-validation parameters as in the paper
```
Comment (timokau) [inline poc/experiment.py:139]: I'm also not sure if I understand the training and validation strategy of ["Leaning Choice  Functions"](https://arxiv.org/pdf/1901.10860.pdf) correctly. I understand that it uses and outer 5-fold cross validation for hyper-parameter optimization, but I'm not sure how the "inner" validation works. From Table 3 I would guess that it works like this:

- Generate (110000 / 4) * 5 = 137,500 instances (to get that 100,000 + 10,000 split in each cross-validation step).
- Split this into 5 folds of 27,500 instances each.
- For each fold F (outer cross-validation loop):
	- Combine the other 4 folds into one dataset of 110,000 instances.
	- Split this into training (100,000) and test (10,000) data
	- For each set of hyperparameters
		- Train a model, test it on the test data.
	- Validate the best (according to test data) model on fold F
- Report the average performance of the outer cross-validation.

However I'm almost sure that is incorrect :smile: Could you clarify @kiudee?

File: poc/experiment.py (hunk 7 above)
Comment (timokau) [reply poc/experiment.py:106]: Okay, thanks! Easy to forget.

**Comment by timokau** [comment]
I have started to remove the tensorflow components. Next steps would include cleanup and integration of the "poc" folder into the main project. We could then merge this PR when it is ready, making the in-progress transition official. Many of the removed components could then be re-introduced / re-written in follow-up PRs.

File: csrank/tests/test_choice_functions.py [hunk 9]
```diff
@@ -1,98 +1,87 @@
...
-    FATELINEAR_CHOICE: (
-        FATELinearChoiceFunction,
-        {},
-        get_vals([0.6558, 0.0722, 0.9998]),
-    ),
-    FETALINEAR_CHOICE: (
-        FETALinearChoiceFunction,
-        {},
-        get_vals([0.8782, 0.8894, 0.9998]),
+        get_vals([0.6861, 0.2014, 1.0]),
```
Comment (timokau) [inline csrank/tests/test_choice_functions.py:57]: The F1 score and the instance informedness went down quite a bit for the FATE general choice model. If I increase the network size slightly (10 instead of 5 for `n_hidden_joint_layers`) the performance goes up again though (`[0.9019, 0.91043, 1.0]`). Most of it is probably just random variation since the example is very small. I think the test is mostly for the basic functionality and the exact numbers do not matter very much here, but I still wanted to highlight the change and start a discussion.

**Comment by timokau** [comment]
I think this is ready for another pair of eyes now. In its current state, the PR

- removes all tensorflow components (while keeping the testsuite and the linters happy),
- adds the "structural components" for pytorch based estimators (using `skorch` as an intermediary) and
- adds a pytorch implementation of FATE scoring and the derived estimators.

It basically "sets the stage" for the remaining estimators. I plan to add FETA, CmpNet and RankNet in follow-up PRs. The focus of this PR is on the structural and architectural aspects. The FATE estimators mostly serve as an example.

Many components that were part of the "proof of concept" at some point are removed now. That includes
- the dataset (the current implementation re-uses the datasets that are already defined in cs-ranking),
- the experiments (these may still be valuable at some point, but are not fleshed out enough to be included in the package in their current state) and
- the FETA implementation.

Other things to note:
- The choice function threshold is currently fixed. Optimizing the threshold is not supported.
- The interactive notebooks were removed. We may want to adapt some of them to the new pytorch estimators at some point, but that is not currently part of this PR.

Open questions:

- To which degree do we want to preserve the commit history? I have already done some rebasing and cleanup. Currently there is one commit that adds the "proof of concept implementations" (a slightly modified version of what I previously developed in this PR) and then a lot of other commits that integrate the pytorch parts into the library. Some parts are restructured, others are removed. For example the experiments and FETA are still in the commit history. We could restructure that a bit to get rid of the "unpolished" intermediate states. We would then have commits like "Add skorch based estimator superclasses", "Add FATE scoring module"). That should not be too much effort. We could also keep it as it is, which adds some context but keeps those unpolished intermediate states.

- At what level of abstraction should the "specialized" estimator classes such as `FATEDiscreteChoiceFunction` sit? Currently they are very low-level, which gives the most flexibility (overriding the scoring module, overriding some aspect of the optimizer etc.) but makes it a bit inconvenient to specify common parameters such as `n_hidden_set_layers`. An alternative would be to add special parameters for those and configure the components accordingly. If somebody wanted full flexibility, they could then use the classes "up the chain" (`SkorchDiscreteChoiceFunction` and `FATEScoring`) to build their own estimator. I think I prefer that alternative now, but both have their benefits. If you agree then we could also make that change in a follow-up PR.

**Comment by timokau** [comment]
I noticed two more things (an issue with the documentation and a difference in test configuration) while working on FETA. I have pushed updates.

Please have a look at the [test configuration commit](https://github.com/kiudee/cs-ranking/pull/164/commits/96111fe4040c664e0d6974ba1e5002106e9e2fa9). I have adjusted it to match the old behavior for now, but we may want to disable validation for tests instead. It is especially odd for the ranking test. I can see one argument in favor of the validation split: The split behavior is at least exercised in the tests. That could catch obvious exception-generating errors. Skorch and its test suite probably does a better job at testing this though.

File: csrank/modules/scoring/fate.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    4.6
+
+    Let's verify this:
+
+    >>> scoring(instances)
+    tensor([[ 4.6000,  6.6000,  8.6000],
+            [16.2000, 18.2000, 20.2000]])
+
+    As you can see, the scoring comes to the same result for the first object
+    of the first instance.
```
Comment (kiudee) [inline csrank/modules/scoring/fate.py:64]: Great documentation!

**Review by kiudee — CHANGES_REQUESTED** [review]
Very good modular structure. The only thing which should be changed is that all hyperparameters are available on the level of the learners (the level which implements the sklearn interface), as we talked about.

File: csrank/modules/scoring/fate.py (hunk 10 above)
Comment (timokau) [reply csrank/modules/scoring/fate.py:64]: Thanks :)

**Comment by timokau** [comment]
> Very good modular structure.

Thanks again :)

> The only thing which should be changed is that all hyperparameters are available on the level of the learners (the level which implements the sklearn interface), as we talked about.

I have just pushed a change that should address this. I agree that the hyperparameters were very inconvenient to configure. It was technically possible, but not as easy as it should be. The new approach corresponds to the alternative I have mentioned in the second open question [here](https://github.com/kiudee/cs-ranking/pull/164#issuecomment-812940505).

File: pyproject.toml [hunk 11]
```diff
@@ -38,6 +38,9 @@ nbsphinx = {version = "^0.7.1", optional = true}
 IPython = {version = "^7.18.1", optional = true}
 torch = "^1.8.0"
 skorch = "^0.9.0"
+# The poetry2nix build of matplotlib fails for newer versions. See
+# https://github.com/nix-community/poetry2nix/issues/280#issuecomment-815064853
+matplotlib = "<3.4"
```
Comment (kiudee) [inline pyproject.toml:43]: Hm, it would be inconvenient, if we have to restrict dependencies just because poetry2nix has problems building it. The dependency specification should only specify a set of dependency ranges for which we know the package works.

File: csrank/choicefunction/fate_choice.py [hunk 12]
```diff
@@ -1,183 +1,77 @@
...
+        n_hidden_set_layers=2,
+        n_hidden_set_units=32,
+        n_hidden_joint_layers=2,
+        n_hidden_joint_units=32,
+        activation=nn.SELU,
+        criterion=nn.BCELoss,
+        **kwargs
+    ):
+        self.n_hidden_set_layers = n_hidden_set_layers
+        self.n_hidden_set_units = n_hidden_joint_units
```
Comment (kiudee) [inline csrank/choicefunction/fate_choice.py:55]: `self.n_hidden_set_units = n_hidden_set_units`

File: csrank/discretechoice/fate_discrete_choice.py [hunk 13]
```diff
@@ -1,136 +1,90 @@
...
-                Arguments to be passed to the Dense layers. See the keras
-                documentation for ``Dense`` for available options.
-        """
-        self.loss_function = loss_function
-        self.metrics = metrics
-        self._store_kwargs(
-            kwargs, {"optimizer__", "kernel_regularizer__", "hidden_dense_layer__"}
-        )
+        self.n_hidden_set_layers = n_hidden_set_layers
+        self.n_hidden_set_units = n_hidden_joint_units
```
Comment (kiudee) [inline csrank/discretechoice/fate_discrete_choice.py:66]: `self.n_hidden_set_units = n_hidden_set_units`

File: csrank/objectranking/fate_object_ranker.py [hunk 14]
```diff
@@ -1,95 +1,85 @@
...
-            n_hidden_joint_units=n_hidden_joint_units,
-            activation=activation,
-            kernel_initializer=kernel_initializer,
-            kernel_regularizer=kernel_regularizer,
-            optimizer=optimizer,
-            batch_size=batch_size,
-            random_state=random_state,
-            **kwargs,
+        self.n_hidden_set_layers = n_hidden_set_layers
+        self.n_hidden_set_units = n_hidden_joint_units
```
Comment (kiudee) [inline csrank/objectranking/fate_object_ranker.py:63]: `self.n_hidden_set_units = n_hidden_set_units` ;)

**Review by kiudee — CHANGES_REQUESTED** [review]
Ready to be merged, after the minor copy paste bug is fixed.

File: csrank/objectranking/fate_object_ranker.py (hunk 14 above)
Comment (timokau) [reply csrank/objectranking/fate_object_ranker.py]: Fixed, thanks for noticing.

**Comment by timokau** [comment]
I have fixed the copy and paste issue that you found.

Just for completeness, I'll summarize the results of our private discussions here too:

- I have removed the poetry2nix specific workaround.
- I have restructured and reduced the history to avoid adding components just to remove them again in the same PR (related to the first open question [here](https://github.com/kiudee/cs-ranking/pull/164#issuecomment-812940505)). Sorry for all the email notifications and force pushes.
- I have "lifted" the level of abstraction of the specialized estimator classes (related to the second open question [here](https://github.com/kiudee/cs-ranking/pull/164#issuecomment-812940505)).
- I have removed the train/validation split in the tests (related to [this](https://github.com/kiudee/cs-ranking/pull/164#issuecomment-814118161) comment).

The PR is ready for review again.

**Review by kiudee — APPROVED** [review]
Great clean commit history. Looks ready to be merged.

**Comment by timokau** [comment]
Thank you for the reviews :)

**Comment by timokau** [comment]
I wanted to run the checks and lints once more before merging and noticed some formatting issues. I did not run `black` consistently because I had a newer version in my environment and that made a lot of unrelated formatting changes. I have fixed the formatting with the `black` version that is defined in the `.pre-commt-config.yaml`. The linters (including `black`) give a thumbs-up now and the test suite passes.

Please take another look.

**Review by kiudee — APPROVED** [review]

**Comment by timokau** [comment]
Thanks again :rocket:
